components/merkledb/examples/migration.rs

```diff
@@ -0,0 +1,260 @@
+        schema
+            .histories
+            .get(&public_key)
+            .extend((0..history_len).map(|_| {
+                let mut bytes = [0_u8; HASH_SIZE];
+                rng.fill(&mut bytes[..]);
+                Hash::new(bytes)
+            }));
+    }
+}
+
+mod v2 {
+    use super::*;
+
+    #[derive(Debug, Serialize, Deserialize)]
+    pub struct Wallet {
+        pub username: String,
+        pub balance: u32,
+        pub history_hash: Hash, // << new field
+    }
+
+    impl BinaryValue for Wallet {
+        fn to_bytes(&self) -> Vec<u8> {
+            bincode::serialize(self).unwrap()
+        }
+
+        fn from_bytes(bytes: Cow<'_, [u8]>) -> Result<Self, Error> {
+            bincode::deserialize(bytes.as_ref()).map_err(From::from)
+        }
+    }
+
+    #[derive(Debug, Serialize, Deserialize)]
+    pub struct Config {
+        pub ticker: String,
+        pub divisibility: u8,
+    }
+
+    impl BinaryValue for Config {
+        fn to_bytes(&self) -> Vec<u8> {
+            bincode::serialize(self).unwrap()
+        }
+
+        fn from_bytes(bytes: Cow<'_, [u8]>) -> Result<Self, Error> {
+            bincode::deserialize(bytes.as_ref()).map_err(From::from)
+        }
+    }
+
+    impl_object_hash_for_binary_value! { Wallet, Config }
+
+    #[derive(Debug, FromAccess)]
+    pub struct Schema<T: Access> {
+        pub config: ProofEntry<T::Base, Config>,
+        pub wallets: ProofMapIndex<T::Base, PublicKey, Wallet>,
+        pub histories: Group<T, PublicKey, ProofListIndex<T::Base, Hash>>,
+    }
+
+    impl<T: Access> Schema<T> {
+        pub fn print_wallets(&self) {
+            for (public_key, wallet) in self.wallets.iter().take(10) {
+                println!("Wallet[{:?}] = {:?}", public_key, wallet);
+                println!(
+                    "History = {:?}",
+                    self.histories.get(&public_key).iter().collect::<Vec<_>>()
+                );
+            }
+        }
+    }
+}
+
+fn migrate(new_data: Prefixed<&Fork>, old_data: Prefixed<ReadonlyFork>) {
+    println!("\nStarted migration");
+    let old_schema = v1::Schema::new(old_data);
+    let mut new_schema = v2::Schema::new(new_data.clone());
+
+    // Move `ticker` and `divisibility` to `config`.
+    let config = v2::Config {
+        ticker: old_schema.ticker.get().unwrap(),
+        divisibility: old_schema.divisibility.get().unwrap_or(0),
+    };
+    new_schema.config.set(config);
+    // Mark these two indexes for removal.
+    new_data.clone().create_tombstone("ticker");
+    new_data.clone().create_tombstone("divisibility");
+
+    // Migrate wallets.
+    for (i, (public_key, wallet)) in old_schema.wallets.iter().enumerate() {
+        if wallet.username == "Eve" {
+            // We don't like Eves 'round these parts. Remove her transaction history
+            // and don't migrate the wallet.
+            new_data
+                .clone()
+                .create_tombstone(("histories", &public_key));
```

Comment:
Do you mean `old_data` or `old_schema`? Using `old_schema.histories` is not possible because it corresponds to the original data and we create a tombstone in the new data.

Reply:
`old_schema`. I`d suggest to add `name` method to indexes and write somethink like code bellow.

```rust
new_data().clone().create_tombstone(old_schema.history.name());
new_data().clone().create_tombstone(old_schema.histories.name());
```

Or even we may introduce trait 
```rust

trait IndexName {
    type Out;
    fn index_name(&self) -> Self::Out
}

impl<_> IndexName for MapIndex<_> { ... }
```

Reply:
But this won't work without additional gizmos: the access provided to the migration script most probably needs to be prefixed, but the name that can be retrieved from an index would not prefixed.

Personally, I don't see a huge problem in having separate UX for altering the schema; e.g., in SQL table altering operations are distinct from data modification operations.